core/src/main/java/org/infinispan/xsite/statetransfer/SyncProviderState.java

```diff
@@ -0,0 +1,239 @@
+                     .concatMapCompletable(this, 1)
+                     .subscribe(this),
+               provider.getExecutor());
+
+      }
+
+      public void cancel() {
+         canceled = true;
+      }
+
+      @Override
+      public boolean test(List<XSiteState> ignored) {
+         //Flowable#takeUntil method
+         return canceled;
+      }
+
+      @Override
+      public void onSubscribe(@NonNull Disposable d) {
+
+      }
+
+      @Override
+      public void onComplete() {
+         //if canceled, the coordinator already cleanup the resources. There is nothing to be done here.
+         if (canceled) {
+            return;
+         }
+         provider.notifyStateTransferEnd(state.getBackup().getSiteName(), coordinator, true);
+         state.taskFinished();
+      }
+
+      @Override
+      public void onError(@NonNull Throwable e) {
+         //if canceled, the coordinator already cleanup the resources. There is nothing to be done here.
+         if (canceled) {
+            return;
+         }
+         provider.notifyStateTransferEnd(state.getBackup().getSiteName(), coordinator, false);
+         state.taskFinished();
+      }
+
+      @Override
+      public CompletableSource apply(List<XSiteState> xSiteStates) {
+         //Flowable#concatMapCompletable method
+         XSiteBackup backup = state.getBackup();
+         //TODO!? can we use xSiteStates directly instead of copying?
+         XSiteState[] privateBuffer = xSiteStates.toArray(new XSiteState[0]);
+
+         if (log.isDebugEnabled()) {
+            log.debugf("Sending chunk to site '%s'. Chunk has %s keys.", backup.getSiteName(), privateBuffer.length);
+         }
+
+         XSiteStatePushCommand command = provider.getCommandsFactory().buildXSiteStatePushCommand(privateBuffer, backup.getTimeout());
+         return Completable.fromCompletionStage(new CommandRetry(backup, command, provider, state.getWaitTimeMillis(), state.getMaxRetries()).send());
+      }
+   }
+
+   private static class CommandRetry extends CompletableFuture<Void> implements java.util.function.BiConsumer<Void, Throwable> {
+
+      private final XSiteBackup backup;
+      private final XSiteStatePushCommand cmd;
+      private final XSiteStateProvider provider;
+      private final long waitTimeMillis;
+      @GuardedBy("this")
+      private int maxRetries;
+
+      private CommandRetry(XSiteBackup backup, XSiteStatePushCommand cmd, XSiteStateProvider provider, long waitTimeMillis, int maxRetries) {
+         this.backup = backup;
+         this.cmd = cmd;
+         this.provider = provider;
+         this.waitTimeMillis = waitTimeMillis;
+         this.maxRetries = maxRetries;
+      }
+
+      //method to invoke from other class
+      CompletionStage<Void> send() {
+         doSend();
+         return this;
+      }
+
+      //used in scheduled executor, invokes doSend after the wait time.
+      void nonBlockingSend() {
+         provider.getExecutor().execute(this::doSend);
```

Comment:
Same thing here about the use of the executor, does the `RpcManager#invokeXSite` method block or something?

Reply:
it can block in JGroups' flow control. And since the `ScheduledExecutorService` (`TIMEOUT_SCHEDULE_EXECUTOR`) only contains a single thread, if the thread blocks you will be unable to trigger timeouts :/